File: ios/SplitBackAPI/Sources/SplitBackAPI/Models/Goal.swift
import Foundation
import SwiftData

/// A budgeting goal, derived from Plaid data. Mirrors the server `goals` table. Two kinds:
/// - `.spend`: a Mint-style monthly budget capping spend in `category` at `targetAmount`.
/// - `.save`: grow `accountId`'s balance; `saveTargetType` is reach-a-balance or save-an-amount,
///   measured from the `startingBalance`/`startingDate` snapshot taken at creation.
/// Progress is computed client-side (see `GoalProgress`) — never stored.
@Model
final class Goal {
    @Attribute(.unique) var id: UUID
    var kind: String
    var name: String
    var category: String?
    var accountId: UUID?
    var targetAmount: Decimal
    var saveTargetType: String?
    var startingBalance: Decimal?
    var startingDate: Date?
    var period: String
    var currency: String
    var archivedAt: Date?
    var createdAt: Date
    var updatedAt: Date

    init(
        id: UUID,
        kind: String,
        name: String,
        category: String? = nil,
        accountId: UUID? = nil,
        targetAmount: Decimal,
        saveTargetType: String? = nil,
        startingBalance: Decimal? = nil,
        startingDate: Date? = nil,
        period: String = "monthly",
        currency: String,
        archivedAt: Date? = nil,
        createdAt: Date,
        updatedAt: Date
    ) {
        self.id = id
        self.kind = kind
        self.name = name
        self.category = category
        self.accountId = accountId
        self.targetAmount = targetAmount
        self.saveTargetType = saveTargetType
        self.startingBalance = startingBalance
        self.startingDate = startingDate
        self.period = period
        self.currency = currency
        self.archivedAt = archivedAt
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    var goalKind: GoalKind { GoalKind(rawValue: kind) ?? .spend }
    var saveTarget: SaveTargetType? { saveTargetType.flatMap(SaveTargetType.init(rawValue:)) }
}

/// What a goal tracks.
enum GoalKind: String, CaseIterable {
    case spend  // a monthly category budget
    case save   // grow an account's balance
}

/// How a savings goal's target is interpreted.
enum SaveTargetType: String, CaseIterable {
    case balance  // reach an absolute balance
    case amount   // add this much from the starting snapshot
}

File: backend/app/models/goal.py
import uuid
from datetime import date as date_type
from datetime import datetime
from decimal import Decimal

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Numeric, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db import Base
from app.models.base import TimestampMixin, UUIDMixin


class Goal(UUIDMixin, TimestampMixin, Base):
    """A budgeting goal derived from Plaid data. Two kinds:
    - "spend": cap monthly spend in a canonical `category` at `target_amount` (Mint-style budget).
    - "save": grow a Plaid `account`'s balance. `save_target_type` is "balance" (reach an absolute
      balance) or "amount" (add this much); progress is measured from the `starting_balance` snapshot
      taken at creation (there is no historical balance series to look back on).
    Progress is computed client-side from cached transactions/accounts — not stored here."""

    __tablename__ = "goals"

    kind: Mapped[str] = mapped_column(String(16), nullable=False)  # "spend" | "save"
    name: Mapped[str] = mapped_column(String(128), nullable=False)
    # The local identifier this goal belongs to (per-caller scoping); the creator. Null = legacy/unowned.
    owner_identifier: Mapped[str | None] = mapped_column(String(128), nullable=True)
    category: Mapped[str | None] = mapped_column(String(64), nullable=True)  # spend goals
    account_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("accounts.id", ondelete="CASCADE"), nullable=True
    )  # save goals
    target_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    save_target_type: Mapped[str | None] = mapped_column(String(16), nullable=True)  # "balance" | "amount"
    starting_balance: Mapped[Decimal | None] = mapped_column(Numeric(12, 2), nullable=True)
    starting_date: Mapped[date_type | None] = mapped_column(Date, nullable=True)
    period: Mapped[str] = mapped_column(String(16), nullable=False, default="monthly")
    currency: Mapped[str] = mapped_column(String(3), nullable=False, default="USD")
    archived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    # Owner-set: when true, the goal is visible (read-only) to the owner's accepted connections.
    shared: Mapped[bool] = mapped_column(Boolean, nullable=False, default=False)
